app/src/main/java/com/neura/sampleapplication/NeuraManager.java

```diff
@@ -46,10 +63,160 @@ public NeuraApiClient getClient() {
         mNeuraApiClient.setAppSecret(context.getResources().getString(R.string.app_secret));
         mNeuraApiClient.connect();
     }
-}
+
+    //create a call back to handle authentication stages.
+    private static AnonymousAuthenticationStateListener silentStateListener = new AnonymousAuthenticationStateListener() {
+        @Override
+        public void onStateChanged(AuthenticationState state) {
+
+            switch (state) {
+                case AccessTokenRequested:
+                    break;
+                case AuthenticatedAnonymously:
+                    // successful authentication
+                    NeuraManager.getInstance().getClient().unregisterAuthStateListener();
+
+//                    NeuraManager.getInstance().getClient().getUserDetails(new UserDetailsCallbacks() {
+//                        @Override
+//                        public void onSuccess(UserDetails userDetails) {
+//                            if (getActivity() != null) {
+//                                userDetails.getData().getNeuraId();
+//                                NeuraManager.getInstance().getClient().getUserAccessToken();
+//                            }
+//                        }
+//
+//                        @Override
+//                        public void onFailure(Bundle resultData, int errorCode) {
+//
+//                        }
+//                    });
+                    break;
+                case NotAuthenticated:
+                case FailedReceivingAccessToken:
+                    // Authentication failed indefinitely. a good opportunity to retry the authentication flow
+                    NeuraManager.getInstance().getClient().unregisterAuthStateListener();
+                    break;
+                default:
+            }
+        }
+    };
+
+    public static void authenticateAnonymously(final FragmentMain context) {
+        if (!isMinVersion()) {
+            return;
+        }
+
+        if (NeuraManager.getInstance().getClient().isLoggedIn()) {
+            return;
+        }
+
+        //Get the FireBase Instance ID, we will use it to instantiate AnonymousAuthenticationRequest
+        String pushToken = FirebaseInstanceId.getInstance().getToken();
+
+        //Instantiate AnonymousAuthenticationRequest instance.
+        AnonymousAuthenticationRequest request = new AnonymousAuthenticationRequest(pushToken);
+
+        //Pass the AnonymousAuthenticationRequest instance and register a call back for success and failure events.
+        NeuraManager.getInstance().getClient().authenticate(request, new AnonymousAuthenticateCallBack() {
+            @Override
+            public void onSuccess(AnonymousAuthenticateData data) {
+                NeuraManager.getInstance().getClient().registerAuthStateListener(silentStateListener);
+                Log.i(TAG, "Successfully requested authentication with neura. ");
+                context.setUIState(true, true);
```

Comment:
it's not understood here what "true, true" means, Android studio helps with this and explains, but it's not a good practice.

Also, from a Neura perspective, this should start a loading state and it will be connected = true only when the listener switch is at "AuthenticatedAnonymously"

Reply:
For the first part, I'll refactor it.

For the second part, it is a bit tricky. In general, having the UI tied into the authentication makes it messy, because getting a final copy of the fragment context in a static state could result in a null context, especially if instance of the fragment is destroyed. Since we don't know when the callback will say the user is authenticated, it would be best to keep the authentication method inside the UI class, or at least in a non-static separate class. Setting connected = true would be safer inside the fragment itself, which would require authenticating within the fragment, too. Let me know if I should still try to keep authenticateAnonymously inside the singleton or move it out.

Reply:
An IntentService might be a good compromise

Reply:
>having the UI tied into the authentication makes it messy
You are right, that's also related to the fact that you pass a fragment to this class.
So, what do you think?
Neura manager should be a *sterile class* you can take from the sample and just put in your project with *no dependencies*.
Every UI/UX choices you make should be in a wrapper

Reply:
I agree, but in this app, the auth callbacks are necessary to update the UI. If I remove and connection to the UI in the auth methods, then the app won't function correctly.

Reply:
I don't agree.
For this app, the status is always "disconnected" until listener is activated (or in the next launch of the app you use the method "isLoggedIn" to change the status.

at most, once tapped for anonymous auth, I'd change disconnected to "Pending" or "Authenticating" and then, when the listener is anonymouslyAuthenticated change the UI. (you can add a listener for this on the UI wrapper for example)